LogAnalysis.hs

```diff
@@ -0,0 +1,101 @@
+{-# OPTIONS_GHC -Wall #-}
+module LogAnalysis where
+import Log
+
+-- 'I' is for informational messages 
+-- 'W' for warnings
+-- 'E' for errors 
+
+-- Error messages have scale on 1 to 100 (i.e. mild to extreme)
+
+---------------------------------------------------------------------------------
+-- Exercise 1: parseMessage
+---------------------------------------------------------------------------------
+parseMessage :: String -> LogMessage  -- parsing function 
+
+-- $ means that anything after it will take precedence over anything before it 
+parseMessage s = case words s!!0 of
+    "I" -> LogMessage Info (read $ words s!!1 :: TimeStamp) (unwords $ drop 2 $ words s)
+    "W" -> LogMessage Warning (read $ words s!!1 :: TimeStamp) (unwords $ drop 2 $ words s)
+    "E" -> LogMessage (Error (read $ words s!!1 :: TimeStamp)) (read $ words s!!2 :: TimeStamp) 
+           (unwords $ drop 3 $ words s) 
+    _   -> Unknown s  -- Easiest to deal with 
+
+parse :: String -> [LogMessage]
+-- Map function takes function and list as arguments, returns list under image of function 
+parse s = map parseMessage $ lines s
+
+---------------------------------------------------------------------------------
+-- Exercise 2: insert
+---------------------------------------------------------------------------------
+-- Raising abstraction with a single function to get the timestamp from a LogMessage first 
+getInt :: LogMessage -> TimeStamp
+getInt (LogMessage _ a _) = a
+getInt (Unknown _) = 0
+getInt (LogMessage (Error _) b _ ) = b
+
+insert :: LogMessage -> MessageTree -> MessageTree
+insert lmX Leaf = Node Leaf lmX Leaf  -- base case, empty node/Leaf 
+insert lmX (Node t lmY s)
+    | getInt lmX <= getInt lmY = Node (insert lmX t) lmY s  -- step case, less than or equal
```

Comment:
Instead of calling `<=` operator (or similar) multiple times, it's more efficient to use `compare` function and then pattern-match on resulting value of `Ordering` type:

* https://hackage.haskell.org/package/base-4.11.1.0/docs/Prelude.html#v:compare